### PR TITLE
feat: add hadolint rules infrastructure and comprehensive tests

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Overview

This PR adds a complete infrastructure for tracking, documenting, and testing Hadolint rules implementation.

### 🏗️ Infrastructure

- **Hadolint rules tracking**: Added `hadolint-status.json` and `hadolint-rules.json` for tracking implementation status
- **Auto-sync workflow**: Weekly cron job (`.github/workflows/sync-hadolint-rules.yml`) to sync rules from upstream Hadolint repo and create PRs automatically
- **Documentation automation**: Scripts to auto-update RULES.md and README.md with current rule counts
- **Pre-push validation**: Hook to ensure implemented rules are documented in status file

### 📦 Rule Modules (Semantic Rules)

Extracted 5 semantic hadolint rules from `internal/semantic/` into individual modules in `internal/rules/hadolint/`:

- `dl3012.go` - Multiple HEALTHCHECK instructions
- `dl3023.go` - COPY --from self-reference detection
- `dl3024.go` - Duplicate stage names detection
- `dl3043.go` - Forbidden ONBUILD triggers
- `dl3061.go` - Invalid instruction order

Each module exports constants, helper functions, and detection logic, making them visible in the filesystem organization.

### ✅ Comprehensive Test Coverage

Created **792 lines** of tests across 5 new test files:

| File | Lines | Test Cases | Coverage |
|------|-------|------------|----------|
| `dl3012_test.go` | 120 | 8 | Multiple HEALTHCHECK detection |
| `dl3023_test.go` | 133 | 10 | Self-referencing COPY detection |
| `dl3024_test.go` | 128 | 8 | Duplicate stage names |
| `dl3043_test.go` | 237 | 47 | Forbidden ONBUILD triggers |
| `dl3061_test.go` | 174 | 35 | Invalid instruction order |
| **Total** | **792** | **108** | All passing ✅ |

All tests systematically extracted from the original Hadolint test suite, with additional edge cases.

### 🛠️ Scripts

- `scripts/extract-hadolint-rules.sh` - Extract rules from Hadolint repo using ast-grep
- `scripts/generate-hadolint-table.sh` - Generate and update RULES.md table + README.md count
- `scripts/validate-hadolint-status.sh` - Validate that implemented rules are documented

### 🔄 CI/CD Improvements

- `.github/workflows/sync-hadolint-rules.yml` - Weekly sync from upstream with auto-PR creation
- Updated `peter-evans/create-pull-request` from v6 to **v8.1.0** (latest)
- Added `validate-hadolint-status` to pre-push hooks in `.lefthook.yml`

### 📚 Documentation

- Added markers to `RULES.md` (`<!-- BEGIN/END HADOLINT_DL_RULES -->`) for auto-updates
- Added markers to `README.md` (`<!-- BEGIN/END RULES_TABLE -->`) for auto-updates
- README now shows: **12 rules** for Hadolint (auto-updated)
- RULES.md hadolint table now auto-generated from status file

## Testing

```bash
# Run all hadolint tests
go test ./internal/rules/hadolint/...

# Validate status file
./scripts/validate-hadolint-status.sh

# Generate and update docs
./scripts/generate-hadolint-table.sh --update
```

All 354 hadolint tests pass, including the 108 new tests for semantic rules.

## Benefits

1. **Visibility**: Semantic rules are now visible in the filesystem at `internal/rules/hadolint/`
2. **Automation**: Weekly sync keeps us up-to-date with Hadolint upstream
3. **Validation**: Pre-push hook catches missing status entries
4. **Documentation**: Auto-updated docs always reflect current state
5. **Testing**: Comprehensive test coverage ensures correctness

## Related

- Addresses the need for better organization of hadolint rules
- Provides infrastructure for future rule additions
- Ensures compatibility with upstream Hadolint

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for multiple additional Hadolint checks (healthcheck counts, self-referencing COPY, duplicate stage names, forbidden ONBUILD triggers, instruction ordering).

* **Documentation**
  * Expanded and linked Hadolint rules with per-rule metadata, refreshed rules table and README summary.

* **Tests**
  * Added unit tests covering the newly supported Hadolint checks.

* **Chores**
  * Weekly workflow to sync Hadolint rules, pre-push validation, and scripts to extract/generate/validate rule metadata.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: scripts/generate-hadolint-table.sh [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+    local merged
+    merged=$(merge_rules "$tmp_rules")
+
+    # Handle update mode
+    if [[ "$do_update" == true ]]; then
+        local table count
+        table=$(echo "$merged" | format_markdown)
+        count=$(echo "$merged" | jq '[.[] | select(.impl_status == "implemented")] | length')
+        update_rules_md "$table"
+        update_readme_md "$count"
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply scripts/generate-hadolint-table.sh]: ✅ Addressed in 468d4de. Thanks @coderabbitai\!
